interval: reduce memory footprint

avikivity opened the pull request:
The interval class's memory footprint isn't important for single objects,
but intervals are frequently held in moderately sized collections. In #3335 this
caused a stall. Therefore reducing interval's memory footprint and reduce
allocation pressure.

This series does this by consolidating badly-padded booleans in the object tree
spanned by interval into 5 booleans that are consecutive in memory. This
reduces the space required by these booleans from 40 bytes to 8 bytes.
 
perf-simple-query report (with refresh-pgo-profiles.sh for each measurement):

before:

252127.60 tps ( 66.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   37128 insns/op,   18147 cycles/op,        0 errors)
INFO  2025-06-07 21:00:34,010 [shard 0:main] group0_tombstone_gc_handler - Setting reconcile time to   1749319231 (min id=4dbed2f4-43c9-11f0-cbc6-87d1a08b4ca4)
246492.37 tps ( 66.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   37153 insns/op,   18411 cycles/op,        0 errors)
253633.11 tps ( 66.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   37127 insns/op,   17941 cycles/op,        0 errors)
254029.93 tps ( 66.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   37155 insns/op,   17951 cycles/op,        0 errors)
254465.76 tps ( 66.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   37123 insns/op,   17906 cycles/op,        0 errors)
throughput:
	mean=   252149.75 standard-deviation=3282.75
	median= 253633.11 median-absolute-deviation=1880.17
	maximum=254465.76 minimum=246492.37
instructions_per_op:
	mean=   37137.24 standard-deviation=15.71
	median= 37127.54 median-absolute-deviation=14.45
	maximum=37155.24 minimum=37122.79
cpu_cycles_per_op:
	mean=   18071.19 standard-deviation=212.25
	median= 17950.62 median-absolute-deviation=130.10
	maximum=18411.50 minimum=17906.13



after:

252561.26 tps ( 66.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   37039 insns/op,   18075 cycles/op,        0 errors)
256876.44 tps ( 66.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   37022 insns/op,   17785 cycles/op,        0 errors)
257084.38 tps ( 66.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   37030 insns/op,   17840 cycles/op,        0 errors)
257305.35 tps ( 66.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   37042 insns/op,   17804 cycles/op,        0 errors)
258088.53 tps ( 66.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   37028 insns/op,   17778 cycles/op,        0 errors)
throughput:
	mean=   256383.19 standard-deviation=2185.22
	median= 257084.38 median-absolute-deviation=922.16
	maximum=258088.53 minimum=252561.26
instructions_per_op:
	mean=   37032.17 standard-deviation=8.06
	median= 37030.46 median-absolute-deviation=6.44
	maximum=37041.83 minimum=37021.93
cpu_cycles_per_op:
	mean=   17856.60 standard-deviation=124.70
	median= 17804.16 median-absolute-deviation=71.24
	maximum=18075.50 minimum=17777.95

A small improvement is observed in instructions_per_op. It could be random fluctuations in the compiler performance, or maybe the default constructor/destructor of interval are meaningful even in this simple test. 



Small performance improvement, so not a backport candidate.

avikivity commented:
v2: drop errant backslash in partition_range_compat.hh, noticed by copilot.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17339/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17339/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17339/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/partitioner_test
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8438/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5269/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16521/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17339/)
#### Build Details:
- Duration: 6 hr 34 min
- Builder: spider8.cloudius-systems.com

avikivity commented:
@tgrabiec ping

avikivity commented:
@tgrabiec ping

tgrabiec commented on interval.hh:
```diff
@@ -88,22 +89,156 @@ public:
+    wrapping_interval& operator=(const wrapping_interval& o) {
+        if (this != &o) {
+            if (_start_exists) {
+                std::destroy_at(&_start_value);
```
Shouldn't we do the same dance for exception safety like in the move-assignment? To avoid double-destroy, and destroy-without-construct in case of exception later in this function?

avikivity replied:
You mean clear _start_exists? yes. I'll add tests.

avikivity replied:
No, we can't throw here. But we do need to improve the code later down.

tgrabiec commented on interval.hh:
```diff
@@ -251,38 +386,78 @@ public:
-            std::swap(_start, _end);
+            std::swap(_start_exists, _end_exists);
+            std::swap(_start_inclusive, _end_inclusive);
+            std::swap(_start_value, _end_value);
```
It's not safe if !_end_exists since _end_value is not initialized. _start_value may also not exist.

avikivity commented:
v4:
 - added exhaustive tests for manual memory management (at least, I was exhausted)
 - fixed copy assignment operator to manage _start_exists/_end_exists tightly with _start_value/_end_value so they don't drift apart and cause the destructor to destroy uninitialized objects
 - simplified copy/move assignment operators not to destroy _start_value on exception; leave it to the destructor
 - made writes to the various booleans unconditional (faster if std::destroy_at() is optimized out)
 - use native algorithm for reverse() to avoid trouble with exception safety
 - removed value-stealing start() and end() overloads, as they are broken for singular intervals
 - rebase
 - added perf-simple-query report

avikivity commented:
I proposed an extension to trivial relocation to make this even better, but hardly anyone understood me. https://lists.isocpp.org/std-proposals/2025/06/14078.php

nyh commented on interval.hh:
```diff
@@ -74,31 +84,166 @@ class wrapping_interval {
+    bool _end_exists = false;
+    bool _end_inclusive = false;
+    bool _singular = false;
+    union {
```
I agree with the AI - you explain in the commit message why you used union and not the traditional optional<>, but a few years from now nobody will understand why you used this 1980s style programming. A comment can help (and what the heck is that "poison" thing?).

But I actually have an even even better idea:

Instead of open-coding the new optional-like feature that uses an external boolean, and doing it twice, you could implement a optional<> like template that takes an external boolean, and then it will be very clear how you use it. 

ChatGPT wrote the following code when I told it what I wanted to do (it took A LOT OF TRIAL AND ERROR - it did a lot of bad versions until I got the following):
```c++
#include <type_traits>
#include <utility>
#include <stdexcept>
#include <new>
#include <cstdint>

template <typename T, std::size_t FlagOffset>
class external_optional {
public:
    external_optional() = default;

    ~external_optional() {
        reset();
    }

    bool has_value() const {
        return *get_flag_ptr();
    }

    T& value() {
        if (!has_value()) throw std::runtime_error("Bad access: no value");
        return *std::launder(reinterpret_cast<T*>(&storage_));
    }

    const T& value() const {
        if (!has_value()) throw std::runtime_error("Bad access: no value");
        return *std::launder(reinterpret_cast<const T*>(&storage_));
    }

    T& operator*() { return value(); }
    const T& operator*() const { return value(); }

    T* operator->() { return &value(); }
    const T* operator->() const { return &value(); }

    void reset() {
        if (has_value()) {
            std::launder(reinterpret_cast<T*>(&storage_))->~T();
            *get_flag_ptr() = false;
        }
    }

    template <typename... Args>
    void emplace(Args&&... args) {
        reset();
        new (&storage_) T(std::forward<Args>(args)...);
        *get_flag_ptr() = true;
    }

private:
    typename std::aligned_storage<sizeof(T), alignof(T)>::type storage_;

    bool* get_flag_ptr() const {
        auto base = reinterpret_cast<std::uintptr_t>(this);
        return reinterpret_cast<bool*>(base + FlagOffset);
    }
};

struct Example {
    bool flag;
    external_optional<std::string, offsetof(Example, flag)> opt;

    Example() : flag(false), opt() {}
};

int main() {
    Example ex;

    ex.opt.emplace("Hello with offset");
    std::cout << "Has value? " << std::boolalpha << ex.opt.has_value() << "\n";
    std::cout << "Value: " << *ex.opt << "\n";

    ex.opt.reset();
    std::cout << "After reset: " << ex.opt.has_value() << "\n";
}
```

That "offset" thing I did (actually chatGPT did after I forced it) to make the implementation zero-overhead (no pointer to flag kept in the external_optional), and also movable (the object containing the flag and the optional can be moved safely).

I'm not sure all of this is actually 100% kosher, I'm guessing you can come up with an even better approach if you think this idea has merit.

avikivity replied:
>     external_optional<std::string, offsetof(Example, flag)> opt;

I don't think this is right - you'd need the offset of flag from opt, not from Example. And I don't think the compiler will let you address opt while defining the type for it.

nyh replied:
My idea (which I told ChatGPT to implement...) was that both the "opt" and the boolean "flag" are members of the same structure, so have a constant offset from each other. I couldn't use an actual pointer to the external flag because: 1. It wastes space, 2. it doesn't work if the containing object is moved.

avikivity replied:
You pass offsetof(Example, flag) but you need offsetof(Example, flag) - offsetof(Example, opt) since your base pointer is &opt.

https://godbolt.org/z/n1cnYTj5T

avikivity replied:
In fact I don't see how your example compiles, since Example is incomplete at the point you compute offsetof(Example, flag).

avikivity replied:
It doesn't compile: https://godbolt.org/z/vz1r8j3j1

nyh replied:
I wouldn't be surprised if it's full of bugs - chatGPT wrote this code, not me ;-)
But anway, maybe this implementation is completely broken, but my main point is that if I understood correctly, in your patch you do exactly the same thing for two optionals - start and end. And I was wondering if it's possible to use some sort of template to do the same thing for both, and possibly also encapsulate what you need to do for both in one place (in one template) instead of in 10 different places througout the code.

As I said, I'm not sure this idea is worth pursuing. It's not like I found any bug in your code, which seems fine. I just hoped there might be a more general solution to this problem. If we had some sort of "external optional" feature, perhaps we could use it in other places as well? But I guess it's not important enough to waste more on your time on.

avikivity replied:
It's not possible as far as I can tell. You need the offset between two members, and that offset is only known after the containing structure is completed.

nyh replied:
Maybe you're right, and yet your code solved a similar problem by hard-coding the boolean's name, right? So I wonder if it can't be template-ized.

I had a chat this time with copilot, and asked it to create something different - a multi_optional that contains several optionals (e.g., start and end) so that the two booleans are adjacent and don't need to be aligned twice. The result is the following. But I'm not sure if it's good enough - because in your case you want one more boolean to be added to the class without more alignment between it and the other booleans. I'm not sure this is doable using inheritence, maybe it is - if the inheritance is not virtual, can the compiler pack the resulting structure better. I'm not sure.

```c++
#include <bitset>
#include <tuple>
#include <type_traits>
#include <utility>
#include <cassert>
#include <new>

template <typename... Ts>
class multi_optional {
    static constexpr size_t N = sizeof...(Ts);
    // Storage for presence bits
    std::bitset<N> present;
    // Raw storage for values
    std::tuple<std::aligned_storage_t<sizeof(Ts), alignof(Ts)>...> storage;

    // Helper to get reference to the value
    template <size_t I>
    using type_at = typename std::tuple_element<I, std::tuple<Ts...>>::type;
    template <size_t I>
    type_at<I>& get_ref() {
        return *reinterpret_cast<type_at<I>*>(&std::get<I>(storage));
    }
    template <size_t I>
    const type_at<I>& get_ref() const {
        return *reinterpret_cast<const type_at<I>*>(&std::get<I>(storage));
    }
public:
    // Default constructor: nothing present
    multi_optional() = default;
    ~multi_optional() { reset_all(); }

    // Set presence and construct value
    template <size_t I, typename... Args>
    void emplace(Args&&... args) {
        reset<I>();
        new (&std::get<I>(storage)) type_at<I>(std::forward<Args>(args)...);
        present.set(I);
    }

    // Reset one value
    template <size_t I>
    void reset() {
        if (present[I]) {
            get_ref<I>().~type_at<I>();
            present.reset(I);
        }
    }

    // Reset all values
    void reset_all() {
        reset_all_impl(std::make_index_sequence<N>());
    }
private:
    template <std::size_t... Is>
    void reset_all_impl(std::index_sequence<Is...>) {
        (reset<Is>(), ...);
    }
public:
    // Check if present
    template <size_t I>
    bool has_value() const { return present[I]; }

    // Get value
    template <size_t I>
    type_at<I>& value() {
        assert(present[I]);
        return get_ref<I>();
    }
    template <size_t I>
    const type_at<I>& value() const {
        assert(present[I]);
        return get_ref<I>();
    }
};


multi_optional<int, double, std::string> mo;
mo.emplace<0>(42);
mo.emplace<2>("hello");

if (mo.has_value<0>()) {
    int x = mo.value<0>();
}
if (mo.has_value<2>()) {
    std::string y = mo.value<2>();
}
```

nyh replied:
I continued playing with the AI (this time with the o1 model) and told it to create a multi_optional_extra template which has multiple optionals where their booleans are next to each other to avoid alignment, plus one extra non-optional member which is placed next to the booleans. You can use that to store the extra boolean (or several things) you need without extra padding.

```c++
#pragma once

#include <array>
#include <tuple>
#include <type_traits>
#include <utility>

template <typename Extra, typename... Ts>
class multi_optional_extra {
private:
    // We store booleans for each optional T.
    std::array<bool, sizeof...(Ts)> engaged_{};

    // The additional non-optional member, placed (in most compilers) right after the booleans.
    // There is still no strict standard guarantee that no extra padding will be introduced.
    Extra extraMember_{};

    // For each T, we store aligned storage.
    using Storage = std::tuple<std::aligned_storage_t<sizeof(Ts), alignof(Ts)>...>;
    Storage storage_{};

    // Helper type alias to retrieve specific element from Ts...
    template <std::size_t I>
    using ElementType = std::tuple_element_t<I, std::tuple<Ts...>>;

    // Access raw storage pointer for the I-th element
    template <std::size_t I>
    ElementType<I>* getPtr() {
        return reinterpret_cast<ElementType<I>*>(&std::get<I>(storage_));
    }

    template <std::size_t I>
    const ElementType<I>* getPtr() const {
        return reinterpret_cast<const ElementType<I>*>(&std::get<I>(storage_));
    }

public:
    // Default constructor does nothing special
    multi_optional_extra() = default;

    // Emplace I-th optional
    template <std::size_t I, typename... Args>
    void emplace(Args&&... args) {
        reset<I>();
        new (static_cast<void*>(getPtr<I>())) ElementType<I>(std::forward<Args>(args)...);
        engaged_[I] = true;
    }

    // Query whether I-th optional is engaged
    template <std::size_t I>
    bool has_value() const {
        return engaged_[I];
    }

    // Access the I-th object (caller must ensure has_value<I>() is true)
    template <std::size_t I>
    ElementType<I>& get() {
        return *getPtr<I>();
    }

    template <std::size_t I>
    const ElementType<I>& get() const {
        return *getPtr<I>();
    }

    // Destroy I-th object if it is engaged
    template <std::size_t I>
    void reset() {
        if (engaged_[I]) {
            getPtr<I>()->~ElementType<I>();
            engaged_[I] = false;
        }
    }

    // Access the extra non-optional member
    Extra& extra() {
        return extraMember_;
    }

    const Extra& extra() const {
        return extraMember_;
    }

    ~multi_optional_extra() {
        // Destroy all engaged objects before destruction
        auto destroyAll = [this]<std::size_t... Is>(std::index_sequence<Is...>) {
            (reset<Is>(), ...);
        };
        destroyAll(std::make_index_sequence<sizeof...(Ts)>{});
    }
};
```

avikivity replied:
This idea can work, but needs lots more effort to implement move/copy construction/assignment.

The "Extra" bit is fragile. If its alignment is one byte, it will fit behind the optionals. If it's higher, padding can be generated.

I think this is premature generalization. We have one case of a structure with two optionals that we need to shrink. I don't think it's worthwhile to develop the generic template and then carry the compile-time burden. It reminds me of in-memory-representation which we reverted in ba7a9d2ac3fcd8076fe19ce9b5b3dc106dbddfd0 which made the same (IMO bad) tradeoff of excessive generalization vs compile time and practicality.

avikivity replied:
> Maybe you're right, and yet your code solved a similar problem by hard-coding the boolean's name, right? So I wonder if it can't be template-ized.

My code works because both the boolean and the T are referenced from the containing structure, and after both are design.

Your first idea fails because you reference the T and the boolean before both are defined, in order to reference the boolean from T's private container.

Your second idea works because you're now again referencing the T and the boolean from a shared container, but the price you pay for it is a mini-language for describing structs so the container can instantiate the bools and the Ts, and any extra members.

The mini-language starts to impose constraints. For example, boost::icl uses an enum to describe the various interval shapes (although, it doesn't directly support unbounded intervals). If we wanted something like that, we'd have to teach the intermediate layer how to translate between enum values and presence/absence of Ts.

nyh commented on interval.hh:
```diff
@@ -88,22 +89,178 @@ public:
+    wrapping_interval(optional<bound> start, optional<bound> end, bool singular = false) {
+#ifdef DEBUG
+        _start_poison = 0xDEADBEEFDEADBEEF;
+        _end_poison = 0xDEADBEEFDEADBEEF;
```
I'm afraid I don't understand why you made such an effort to do this "poison" thing. It seems you don't ever assert that poison is still set to that value, so what's the point? Also, it seems your code has a post-condition that always when the flag is true, the value will NOT be set to this poison. So why do we even need this poison feature - just to check for bugs in this optional<>-like code itself?

avikivity replied:
I'll document it. The idea is that after poisoning, referencing the value incorrectly (say via the destructor) will crash. Otherwise, if it's zero (after moving a unique_ptr away, used in debug mode), a bad destroy doesn't trap.

avikivity commented on idl/range.idl.hh:
```diff
@@ -14,14 +14,14 @@ class interval_bound {
 class wrapping_interval {
-    std::optional<interval_bound<T>> start_ref();
-    std::optional<interval_bound<T>> end_ref();
+    std::optional<interval_bound<T>> start();
```
The rename allowed me to audit all calls and allows a reviewer to verify my audit.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17685/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17685/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17685/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/partitioner_test
:small_blue_diamond: boost/wrapping_interval_test
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16782/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5523/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8700/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17685/)
#### Failed Tests (1/45430):
- [test_kill_coordinator_during_op.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17685/testReport/junit/cluster/test_crash_coordinator_before_streaming/Tests___Unit_Tests___test_kill_coordinator_during_op_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op.debug.1)
#### Build Details:
- Duration: 13 hr
- Builder: spider9.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17685/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17685/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17685/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/partitioner_test
:small_blue_diamond: boost/wrapping_interval_test
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16782/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5523/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8700/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17685/)
#### Build Details:
- Duration: 16 hr
- Builder: spider9.cloudius-systems.com

nyh commented:
As usual, CI failed for no apparent reason.

I have to run it again, because I have no idea how much was even tested (it's not like a case where I know that just a single flaky test failed - here I have no idea what was even tested). So I'll restart the CI.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17690/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17690/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17690/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/partitioner_test
:small_blue_diamond: boost/wrapping_interval_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5527/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8704/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16786/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17690/)
#### Failed Tests (1/45425):
- [test_random_failures[stop_after_starting_group0_service-add_new_udt].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17690/testReport/junit/cluster.random_failures/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_after_starting_group0_service_add_new_udt__debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_starting_group0_service-add_new_udt].debug.3)
#### Build Details:
- Duration: 7 hr 22 min
- Builder: spider2.cloudius-systems.com

avikivity commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17690/) ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17690/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17690/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/partitioner_test 🔹 boost/wrapping_interval_test ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5527/) ✅ - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8704/) ✅ - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16786/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17690/)
> 
> #### Failed Tests (1/45425):
> * [test_random_failures[stop_after_starting_group0_service-add_new_udt].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17690/testReport/junit/cluster.random_failures/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_after_starting_group0_service_add_new_udt__debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures%5Bstop_after_starting_group0_service-add_new_udt%5D.debug.3)
> 
> #### Build Details:
> * Duration: 7 hr 22 min
> * Builder: spider2.cloudius-systems.com

https://github.com/scylladb/scylladb/issues/24513

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17695/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17695/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17695/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/partitioner_test
:small_blue_diamond: boost/wrapping_interval_test
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16789/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5530/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8707/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17695/)
#### Build Details:
- Duration: 6 hr 23 min
- Builder: spider2.cloudius-systems.com
